sbncode/EventGenerator/MultiPart/MultiPartRain_module.cc

```diff
@@ -336,30 +354,50 @@ void MultiPartRain::GenPosition(double& x, double& y, double& z) {
   z = fFlatRandom->fire(_zrange[0],_zrange[1]);
 
   if(_debug>0) {
-    std::cout << "Generating a rain particle at (" 
+    std::cout << "Generating a rain particle at ("
 	      << x << "," << y << "," << z << ")" << std::endl;
   }
 }
 
 void MultiPartRain::GenMomentum(const PartGenParam& param, const double& mass, double& px, double& py, double& pz,
 				const double x, const double y, const double z) {
-    
+
     double tot_energy = 0;
     if(param.use_mom)
       tot_energy = sqrt(pow(fFlatRandom->fire(param.kerange[0],param.kerange[1]),2) + pow(mass,2));
     else
       tot_energy = fFlatRandom->fire(param.kerange[0],param.kerange[1]) + mass;
-    
+
     double mom_mag = sqrt(pow(tot_energy,2) - pow(mass,2));
 
+		/* Generating unit vector with uniform distribution
+		 * in direction = over the sphere.
+		 *
+		 * It is sufficient to draw a normal variable in
+		 * each direction and normalize.
+		 *
+		 * https://mathworld.wolfram.com/SpherePointPicking.html
+		 */
+
     if(!param.direct_inward) {
       std::cout<<"No inward directioning..."<<std::endl;
-      double phi   = fFlatRandom->fire(0, 2 * 3.141592653589793238);
-      double theta = fFlatRandom->fire(0, 1 * 3.141592653589793238);
-    
-      px = cos(phi) * sin(theta);
-      py = sin(phi) * sin(theta);
-      pz = cos(theta);
+
+			if (_cosmic_distribution) {
+				double phi   = fFlatRandom->fire(0, 2 * 3.141592653589793238);
```

Comment:
Or ([`#include "larcoreobj/SimpleTypesAndConstants/PhysicalConstants.h"`](https://nusoft.fnal.gov/larsoft/doxsvn/html/namespaceutil.html#aa522ec41346933950a2a529e5bc0f47f)):
```suggestion
				double phi   = fFlatRandom->fire(0, 2 * util::pi());
```